[crmsh-4.6] Dev: pre-migration: implement pre-migration checks for SLES 16 (jsc#PED-11808)

nicholasyang2022 opened the pull request:
This pull request implements pre-migration checks for SLES 16. These checks are expected to run before migrating to SLES 16.

These checks ensure all the cluster nodes runs the latest version of corosync 2.x and pacemaker 2.1.x, and report used resource agents and fence agents that will be removed in SLES 16. It also provides advice for further actions need to take after migration.

---

# Use Cases

## `crm cluster health sles16`

```
suse@ha-1-1:~> sudo crm cluster health sles16
------ corosync @ ha-1-2 ------
[WARN] Cluster services are not running. Check results may be outdated or inaccurate.
       * corosync
       * pacemaker
[WARN] Corosync transport "udpu" will be deprecated in corosync 3. Please use "knet".

------ corosync @ localhost ------
[INFO] Checking dependency version...
[FAIL] Pacemaker version not supported
       Supported version: 2.1.7 <= Pacemaker < 3
       Actual version:    Pacemaker == 2.1.5
[INFO] Checking service status...
[INFO] Checking used corosync features...
[WARN] Corosync transport "udpu" will be deprecated in corosync 3. Please use "knet".
------ cib ------
[INFO] Checking used resource agents...
[FAIL] SAPHanaSR Classic will be removed in SLES 16.
       Before migrating to SLES 16, replace it with SAPHanaSR-angi.
[WARN] stonith:external/sbd will be removed. Please use stonith:fence_sbd

****** summary ******
[FAIL] The pacemaker cluster stack can not migrate to SLES 16. Please fix all the "FAIL" problems above before migrating to SLES 16.
```

## `crm cluster health sles16 --fix`

```
> sudo crm cluster health sles16 --fix
ERROR: "--fix" is only available in SLES 16.
```

## When some of the nodes are offline

```
------ corosync @ ha-1-2 ------
Cannot create SSH connection to root@ha-1-2: ssh: connect to host ha-1-2 port 22: No route to host

------ corosync @ localhost ------
[INFO] Checking dependency version...
[INFO] Checking service status...
[INFO] Checking used corosync features...
[WARN] Corosync transport "udpu" will be deprecated in corosync 3. Please use "knet".
------ cib ------
[INFO] Checking used resource agents...

****** summary ******
[FAIL] The pacemaker cluster stack can not migrate to SLES 16. Please fix all the "FAIL" problems above before migrating to SLES 16.
```

liangxin1300 commented on crmsh/migration-supported-resource-agents.txt:
```diff
@@ -0,0 +1,144 @@
+stonith:fence_vmware-rest
+stonith:fence_wti
+stonith:fence_xenapi
+stonith:fence_zvm
```
How has the above data been generated?
Can we automatically do that based on the local installed resource-agents and fence-agents-* rpm?

nicholasyang2022 replied:
> How has the above data been generated?

Extracted the file list from resource-agents/fence-agents.

> Can we automatically do that based on the locally installed resource-agents and fence-agents-* rpm?

No, because we need to know what is supported by SLES 16 instead of SLES 15 SP6.

liangxin1300 commented:
Suggestion: add completer for `crm cluster health`, now `hawk2` can be completed, so do `sles16`

liangxin1300 commented:
Suggestion:
#### 6. How about there are Failed Resource Actions?
For some reason, some RA failed. For now, `crm cluster health sles16` seems quiet for such an error.
I think we should hint the user to address that failed RA, then do the migration job

liangxin1300 commented on crmsh/migration.py:
```diff
@@ -0,0 +1,379 @@
+    handler.log_info("Checking used corosync features...")
+    transport = 'udpu' if corosync.is_unicast() else 'udp'
+    handler.handle_tip(f'Corosync transport "{transport}" will be deprecated in corosync 3.', [
+        'After migrating to SLES 16, run "crm health sles16 --fix" to migrate it to transport "knet".',
```
Should be `crm cluster health sles16 --fix`

liangxin1300 commented on crmsh/migration.py:
```diff
@@ -0,0 +1,379 @@
+    ])
+    if corosync.get_value("totem.rrp_mode") in {'active', 'passive'}:
+        handler.handle_tip(f'Corosync RRP will be deprecated in corosync 3.', [
+            'After migrating to SLES 16, run "crm health sles16 --fix" to migrate it to knet multilink.',
```
Should be `crm cluster health sles16 --fix`

nicholasyang2022 commented:
> 1. Add completer for crm cluster health, now hawk2 can be completed, so do sles16

:heavy_check_mark:

> 2. Add success info
> 3. Add failure error

:heavy_check_mark: in fd27f8ba8b9012b73793c1321225c09c4c6dec0

> I prefer this [INFO] one, since [PASS] is not the log level

> 4. Add above output into /var/log/crmsh/crmsh.log log file, too

No. These are not logs. They are user interface.

> And, change [FAIL] to [ERROR]

No. This breaks text alignment in the UI.

> 5. Keep the color of node line consistency

:heavy_check_mark: in 9f8e35e5117aefb36576fd4be640d9780ceb7673

> 6. How about there are Failed Resource Actions?

This looks more like a general health check instead of a migration check. Let's discuss further before working the code.

liangxin1300 commented:
Suggestion
#### 10. Add some behave functional test cases to hit the coverage

nicholasyang2022 commented:
> 7. crm report collect pre-migration info

Let's keep this pull request to focus on implement pre-migration checks, and add it in a separated pull request.

> 8. Inaccurate usage?

Fixed.

> 9. Fence agents checking

Added a message specially for `stonith:external/sbd` in 6075723cc2b8f57666650a307aad194bb44a6494.

liangxin1300 commented:
> > 7. crm report collect pre-migration info
> 
> Let's keep this pull request to focus on implement pre-migration checks, and add it in a separated pull request.

I will create one PR about that

liangxin1300 commented on crmsh/cibquery.py:
```diff
@@ -0,0 +1,25 @@
+"""utilities for parsing CIB xml"""
```
Maybe it's better to move these codes into xmlutil.py instead of creating a new py  file?

nicholasyang2022 replied:
No. xmlutil.py is too long, and its name does not reflect its content. Instead of general xml processing utilities, it is a mixed mess with shell utils, CIB specific routines, and even CLI handlers.

liangxin1300 commented on crmsh/cibquery.py:
```diff
@@ -0,0 +1,25 @@
+    )
+
+
+def has_primitive_filesystem_ocfs2(cib: lxml.etree.Element) -> bool:
```
Considering we might use this to detect gfs2, could you please change 'ocfs2' as a parameter?

liangxin1300 commented on crmsh/migration.py:
```diff
@@ -0,0 +1,377 @@
+
+def check_unsupported_resource_agents(handler: CheckResultHandler):
+    handler.log_info("Checking used resource agents...")
+    crm_mon = xmlutil.CrmMonXmlParser()
```
This is not used

liangxin1300 commented on crmsh/migration-supported-resource-agents.txt:
How about moving this file under `crmsh/crmsh/data`, in case there are other txt files in the future?

nicholasyang2022 replied:
I prefer not to do that. I think this will need to add a python module `crmsh.data`, making our build scripts more complex.

liangxin1300 commented on crmsh/migration.py:
```diff
@@ -0,0 +1,377 @@
+            sys.stdout.write(' Good to migrate to SLES 16.\n\n')
+
+
+def check(args: typing.Sequence[str]) -> int:
```
It is better to add a docstring to:
- explain what the `--json` and `--local` option are used for
- introduce this function and return code

nicholasyang2022 replied:
This is quite self-explained. Adding docstrings like "migration.check: run migration checks", or "--json    output the result in json format" does not make sense. And the return code is also the common main() style return code.

liangxin1300 commented:
This usage is correct
```
# crm cluster health xxx
usage: crm cluster health [-h] [-f] {hawk2,sles16}
crm cluster health: error: argument component: invalid choice: 'xxx' (choose from 'hawk2', 'sles16')
```
But not this
```
# crm cluster health sles16 fs
usage: crm [-h] [--json [{oneline,pretty}]] [--local]
crm: error: unrecognized arguments: fs
```

nicholasyang2022 commented:
Usages changed to

```
suse@ha-1-1:~> sudo crm cluster health foo
usage: health [-h] [-f] {hawk2,sles16}
health: error: argument component: invalid choice: 'foo' (choose from 'hawk2', 'sles16')
```

```
suse@ha-1-1:~> sudo crm cluster health sles16 foo
usage: sles16 [-h] [--json [{oneline,pretty}]] [--local]
sles16: error: unrecognized arguments: foo
```

liangxin1300 commented on doc/crm.8.adoc:
```diff
@@ -1006,14 +1006,30 @@ See "crm cluster help geo_join" or "crm cluster geo_join --help"
+* `sles16`: check whether the cluster is good to migrate to SLES 16.
+
+The optional `--fix` argument attempts to automatically resolve any detected
+issues.
```
Please add note that --fix only available at SLES 16

nicholasyang2022 commented on crmsh/migration.py:
```diff
@@ -0,0 +1,376 @@
+    handler.handle_tip(f'Corosync transport "{transport}" will be deprecated in corosync 3.', [
+        'After migrating to SLES 16, run "crm cluster health sles16 --fix" to migrate it to transport "knet".',
+    ])
+    if corosync.get_value("totem.rrp_mode") in {'active', 'passive'}:
```
`none` means SRP, which is supported and does not needs any actions.

liangxin1300 reviewed (approved):
Looks good to me, great work!

Thanks!

zzhou1 reviewed (changes requested):
Great work, some feedback,

- Reporting the CIB related check result in "localhost" is misleading, since they are concepts at the cluster level when the cluster is running. 

- What to interpret the cluster is down? Since in this case, the CIB check is kind of local, the open question is should CIB check run in the remote node?

- "Good to migrate to SLES 16." on ha-1-2 is not right, while CIB check is "FAIL" at the cluster level in the summary section.

zzhou1 commented:
Wonderful, thanks for the great work! 
Let's merge this PR into the development branch instead of the official 4.6 branch for the timebeing.
